### PR TITLE
Refactor `doGenerate` in `C` generator and parts of Python and Typescript docker generators

### PR DESCRIPTION
**Pull request by housengw**
Changes:
- [x] Use local variable in `doGenerate` instead of changing `topLevelName`
- [x] Get rid of class variable `topLevelName` 
- [x] Factored out `runCommand` global variable from `CGenerator`
- [x] clean up unused imports
- [x] Move all docker file related functions and variables away from `GeneratorBase`, `CGenerator`, and `PythonGenerator` into a dedicated code generator class (`DockerGeneratorBase.java`)

### CODE REVIEW

**Comment by housengw** [comment]
Currently `fileConfig` is reassigned to an instance of `FedFileConfig` inside `CGenerator.doGenerate` and `PythonGenerator.doGenerate`. This reflects the assumption that `fileConfig.srcGenPath` can point to different locations at different places of `doGenerate`, which is a bad design. I took several attempts at getting rid of this design pattern but failed because of the ubiquity of code that depend on that assumption (mostly in federated code, ex. `CCompiler`, `CCmakeGenerator`, ...) as well as my lack of understanding of the compile process. I have moved the reassignment of `fileConfig` inside `CGenerator.setUpFederateSpecificParameters`.

**Comment by Soroosh129** [comment]
The reasoning behind this reassignment is simple: we want each federate to have its own src-gen folder. If you get rid of this overriding, all federates will be put in a single folder.

The latter used to be the original behavior for federated programs and this overriding of fileConfig was a patchwork to get the new (folderalized?) behavior.

The overwritten fileConfig should only be effective inside doGenerate in CGenerator within the `for(FederateInstance federate:federates)` loop. I think there are only a handful of things that need fileConfig there. For example, copying necessary files to the src-gen folder needs to know that each federate has a separate folder, etc.

**Comment by lhstrh** [comment]
I agree with @housengw that reassigning the `fileConfig` and `topLevelName` class variable is is a bad practice that makes the code base unnecessarily brittle. There are other ways of configuring the output behavior of code generators that are more modular.

**Comment by Soroosh129** [comment]
> I agree with @housengw that reassigning the fileConfig and topLevelName class variable is is a bad practice

Right. As I said in my previous comment, there should only be a handful of places that are affected by the fileConfig reassignment, so I would expect it to be relatively straightforward to fix.

**Comment by lhstrh** [comment]
I got the sense that @housengw could use some help with that. Would you be able to take a stab at it?

**Comment by Soroosh129** [comment]
> I got the sense that @housengw could use some help with that. Would you be able to take a stab at it?

Let’s see if my comment above helps @housengw first :)

**Comment by housengw** [comment]
I think this PR has reached a good stopping point. I will address @Soroosh129's comment in a future PR.

File: org.lflang/src/org/lflang/generator/DockerGeneratorBase.java [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+        this.isFederated = isFederated;
+        nFederates = 0;
+    }
+
+    public void setHost(String host) {
+        this.host = host;
+    }
+
+    /**
+     * Adds a federate to the list of federates to generate docker files for.
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/DockerGeneratorBase.java:69]:
```suggestion
     * Add a federate to the list of federates to generate docker files for.
```

File: org.lflang/src/org/lflang/generator/c/CDockerGenerator.java [hunk 2]
```diff
@@ -1,42 +1,66 @@
...
+    private final String defaultBaseImage = "alpine:latest";
+
+    public CDockerGenerator(boolean isFederated, boolean CCppMode, TargetConfig targetConfig) {
+        super(isFederated);
+        this.CCppMode = CCppMode;
+        this.targetConfig = targetConfig;
+    }
+
+    /**
+     * Generates the contents of the docker file.
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/c/CDockerGenerator.java:26]:
```suggestion
     * Generate the contents of the docker file.
```

File: org.lflang/src/org/lflang/generator/c/CGenerator.java [hunk 3]
```diff
@@ -2376,7 +2235,47 @@ protected void setUpParameters(LFGeneratorContext context) {
             pickScheduler();
         }
         pickCompilePlatform();
-        parseTargetParameters();
+    }
+
+    // Perform set up that does not generate code
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/c/CGenerator.java:2240]: Javadoc missing

File: org.lflang/src/org/lflang/generator/c/CMainGenerator.java [hunk 4]
```diff
@@ -1,11 +1,75 @@
...
+    }
+
+    private String generateMainFunction() {
         return String.join("\n",
             "int main(int argc, char* argv[]) {",
             "    return lf_reactor_c_main(argc, argv);",
             "}"
         );
     }
+
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/c/CMainGenerator.java:34]: Javadoc missing.

File: org.lflang/src/org/lflang/generator/c/CMainGenerator.java [hunk 5]
```diff
@@ -1,11 +1,75 @@
...
+        parseTargetParameters();
+    }
+
+    public String generateCode() {
+        CodeBuilder code = new CodeBuilder();
+        code.pr(generateMainFunction());
+        code.pr(generateSetDefaultCliOption());
+        return code.toString();
+    }
+
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/c/CMainGenerator.java:26]: Javadoc missing.

File: org.lflang/src/org/lflang/generator/c/CMainGenerator.java [hunk 6]
```diff
@@ -1,11 +1,75 @@
...
+    private TargetConfig targetConfig;
+    /** The command to run the generated code if specified in the target directive. */
+    private List<String> runCommand;
+
+    public CMainGenerator(TargetConfig targetConfig) {
+        this.targetConfig = targetConfig;
+        runCommand = new ArrayList<>();
+        parseTargetParameters();
+    }
+
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/c/CMainGenerator.java:19]: Javadoc misisng.

File: org.lflang/src/org/lflang/generator/DockerGeneratorBase.java [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+     *
+     * @author{Hou Seng Wong <housengw@berkeley.edu>}
+     */
+    protected enum Key {
+        DOCKER_FILE_PATH,
+        DOCKER_FILE_CONTENT,
+        DOCKER_COMPOSE_SERVICE_NAME,
+        DOCKER_BUILD_CONTEXT,
+    }
+
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/DockerGeneratorBase.java:56]: Javadoc missing.

File: org.lflang/src/org/lflang/generator/DockerGeneratorBase.java [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+    public DockerGeneratorBase(boolean isFederated) {
+        moduleNameToData = new HashMap<>();
+        composeServices = new StringBuilder();
+        this.isFederated = isFederated;
+        nFederates = 0;
+    }
+
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/DockerGeneratorBase.java:63]: Javadoc missing.

File: org.lflang/src/org/lflang/generator/python/PythonDockerGenerator.java [hunk 9]
```diff
@@ -1,18 +1,39 @@
...
-    public static String generateDockerFileContent(String topLevelName, Path srcGenPath) {
-        return String.join("\n", 
-            "# Generated docker file for "+topLevelName+".lf in "+srcGenPath+".",
+    public PythonDockerGenerator(boolean isFederated, TargetConfig targetConfig) {
+        super(isFederated);
+        this.targetConfig = targetConfig;
+    }
+
+    /**
+     * Generates the contents of the docker file.
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/python/PythonDockerGenerator.java:20]:
```suggestion
     * Generate the contents of the docker file.
```

File: org.lflang/src/org/lflang/generator/python/PythonGenerator.java [hunk 10]
```diff
@@ -875,6 +862,14 @@ private static String generateMacroEntry(String key, String val) {
         return "(" + StringUtil.addDoubleQuotes(key) + ", " + StringUtil.addDoubleQuotes(val) + ")";
     }
 
+    private static String generatePythonModuleName(String lfModuleName) {
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/python/PythonGenerator.java:865]: Java doc missing (also the next method).

File: org.lflang/src/org/lflang/generator/ts/TSDockerGenerator.kt [hunk 11]
```diff
@@ -20,22 +22,4 @@ class TSDockerGenerator (
...
-        |        build:
-        |            context: .
-        |            dockerfile: HelloWorldContainerized.Dockerfile
-        |networks:
-        |    default:
-        |        name: lf
-        """
-        return dockerComposeFileContent.trimMargin()
-    }
 }
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/ts/TSDockerGenerator.kt:25]: Newline missing.

**Review by lhstrh — APPROVED** [review]
This looks great! Another step into the right direction :-) I only had some minor nitpicks regarding comments.

**Review by lhstrh — APPROVED** [review]
:rocket::rocket::rocket:

**Pull request merged**